### PR TITLE
refactor(op-acceptor): test runner

### PR DESCRIPTION
**Pull request by scharissis**
Refactors the monolithic 1,755-line runner/runner.go file into 5 focused components to improve maintainability and separation of concerns:

  - TestExecutor: Individual test execution and process management
  - OutputParser: Go test JSON output parsing with timeout handling
  - ResultCollector: Hierarchical result aggregation (gates/suites/tests)
  - TestCoordinator: Test workflow orchestration
  - JSONStore: Raw JSON storage and retrieval

### CODE REVIEW

File: op-acceptor/runner/adapters.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+// RunParallel adapts the existing ExecuteTests method to the interface
+func (a *parallelRunnerAdapter) RunParallel(ctx context.Context, validators []types.ValidatorMetadata, result *RunnerResult) error {
+	// Convert validators to work items
+	var workItems []TestWork
+	for _, validator := range validators {
+		gateName := validator.Gate
+		if gateName == "" {
+			gateName = "default"
```
Comment (serpixel) [inline op-acceptor/runner/adapters.go:31]: Is it possible that we will not have a gate defined? Is it for in-memory tests?

File: op-acceptor/runner/collector.go [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+	}
+	// Get or create gate
+	gate, exists := result.Gates[gateName]
+	if !exists {
+		gate = &GateResult{
+			ID:          gateName,
+			Description: gateName,
+			Tests:       make(map[string]*types.TestResult),
+			Suites:      make(map[string]*SuiteResult),
+			Status:      types.TestStatusPass,
```
Comment (serpixel) [inline op-acceptor/runner/collector.go]: Why are we adding by default the test result as passing?

File: op-acceptor/runner/collector.go [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+		}
+		if gate.Status == types.TestStatusFail {
+			anyFailed = true
+		}
+	}
+
+	return c.determineStatusFromFlags(allSkipped, anyFailed)
+}
+
+func (c *resultCollector) determineStatusFromFlags(allSkipped, anyFailed bool) types.TestStatus {
```
Comment (serpixel) [inline op-acceptor/runner/collector.go]: I am wondering if we require these methods to be fair, seems like we could simplify this

File: op-acceptor/runner/parser.go [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+	// Set error message if test failed
+	if result.Status == types.TestStatusFail && errorMsg.Len() > 0 {
+		result.Error = fmt.Errorf("%s", strings.TrimSpace(errorMsg.String()))
+	}
+
+	// Handle skip status
+	if hasSkip && result.Status != types.TestStatusFail {
+		result.Status = types.TestStatusSkip
+		if errorMsg.Len() > 0 && result.Error == nil {
+			result.Error = fmt.Errorf("SKIP: %s", strings.TrimSpace(errorMsg.String()))
```
Comment (serpixel) [inline op-acceptor/runner/parser.go]: Why are we labelling a skip as an error? Perhaps we just should add it as an info message

File: op-acceptor/runner/adapters.go (hunk 1 above)
Comment (scharissis) [reply op-acceptor/runner/adapters.go:31]: Yes.
There's a new "ageless" mode

File: op-acceptor/runner/collector.go (hunk 2 above)
Comment (scharissis) [reply op-acceptor/runner/collector.go]: Optimism :)

For simplicity, I didn't want to introduce an 'Unknown' status, which meant this default either had to be Pass,Skip,Fail.
It then gets re-calculated later in `determineSuiteStatus()` and `determineGateStatus()`.

In the interest of time, I'm going to assume that you're concerned that it may result in false positives. I've changed the default to FAIL and added some tests, including to cover the edge case in which we have no tests in a gate.

File: op-acceptor/runner/parser.go (hunk 4 above)
Comment (scharissis) [reply op-acceptor/runner/parser.go]: Good spot, this was a mistake. I've corrected this.

File: op-acceptor/runner/collector.go (hunk 3 above)
Comment (scharissis) [reply op-acceptor/runner/collector.go]: I'm not sure what you mean by these methods being "fair".

I've gone ahead and simplified things by reusing the one package-level implementation.

**Review by serpixel — APPROVED** [review]

File: op-acceptor/runner/adapters.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	}
+
+	// Copy results from executor result to the provided result
+	copyResults(executorResult, result)
+	return nil
+}
+
+// SetConcurrency sets the concurrency level (this is handled at executor creation time)
+func (a *parallelRunnerAdapter) SetConcurrency(concurrency int) {
+	// The existing ParallelExecutor doesn't support runtime concurrency changes
```
Comment (janjakubnanista) [inline op-acceptor/runner/adapters.go]: Maybe `panic` or at least log here?

File: op-acceptor/runner/collector.go [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+
+	// Update statistics for a test result
+	UpdateStats(result *RunnerResult, gate *GateResult, suite *SuiteResult, test *types.TestResult)
+
+	// Finalize results and calculate statuses
+	FinalizeResults(result *RunnerResult)
+}
+
+// resultCollector implements ResultCollector
+type resultCollector struct{}
```
Comment (janjakubnanista) [inline op-acceptor/runner/collector.go:27]: I'd add a line here to verify that this really implements `ResultsCollector`:

```go
var _ ResultCollector = (*ResultCollector)(nil)
```

File: op-acceptor/runner/executor.go [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+type TestExecutor interface {
+	// Execute runs a single test or package based on the metadata provided.
+	// If metadata.FuncName is empty, it will execute all tests in the package.
+	Execute(ctx context.Context, metadata types.ValidatorMetadata) (*types.TestResult, error)
+
+	// ExecutePackage runs all tests in a specific package.
+	ExecutePackage(ctx context.Context, metadata types.ValidatorMetadata) (*types.TestResult, error)
+}
+
+// testExecutor implements TestExecutor
```
Comment (janjakubnanista) [inline op-acceptor/runner/executor.go:29]: Same as above - add that line that confirms it really implements it

File: op-acceptor/runner/executor.go [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+}
+
+// NewTestExecutor creates a new test executor
+func NewTestExecutor(testDir string, timeout time.Duration, goBinary string, envProvider func() Env,
+	cmdBuilder func(ctx context.Context, name string, arg ...string) (*exec.Cmd, func()),
+	outputParser OutputParser, jsonStore JSONStore) TestExecutor {
+
+	// Input validation
+	if testDir == "" {
+		panic("testDir cannot be empty")
```
Comment (janjakubnanista) [inline op-acceptor/runner/executor.go]: Are we sure we want to `panic` here? Is it too difficult to return and handle errors instead?

File: op-acceptor/runner/jsonstore.go [hunk 9]
```diff
@@ -0,0 +1,64 @@
+package runner
+
+import (
+	"sync"
+
+	"github.com/ethereum-optimism/infra/op-acceptor/logging"
+)
+
+// jsonStore implements JSONStore interface using FileLogger's RawJSONSink
+type jsonStore struct {
```
Comment (janjakubnanista) [inline op-acceptor/runner/jsonstore.go:12]: Same as above

File: op-acceptor/runner/parser.go [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/ethereum-optimism/infra/op-acceptor/types"
+)
+
+// outputParser implements OutputParser interface
+type outputParser struct{}
```
Comment (janjakubnanista) [inline op-acceptor/runner/parser.go:17]: Same as above

File: op-acceptor/runner/parser.go [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+	for scanner.Scan() {
+		line := scanner.Bytes()
+		event, err := parseTestEvent(line)
+		if err != nil {
+			continue
+		}
+
+		if isMainTestEvent(event, metadata.FuncName) {
+			processMainTestEvent(event, result, &testStart, &testEnd, &errorMsg, &hasSkip)
+		} else if event.Test != "" && (strings.Contains(event.Test, "/") || metadata.FuncName == "") {
```
Comment (janjakubnanista) [inline op-acceptor/runner/parser.go]: I think it would add tons of readability if this `else` was also pulled out into a well named function, as it is I am struggling to understand what it means (as opposed to `isMainTestEvent` which is nice and understandable)

File: op-acceptor/runner/runner_test.go [hunk 12]
```diff
@@ -1732,7 +1732,7 @@ gates:
 	require.Contains(t, result.SubTests, "TestSkipped", "Skipped test should be present in results")
 	skippedTest := result.SubTests["TestSkipped"]
 	require.Equal(t, types.TestStatusSkip, skippedTest.Status, "TestSkipped should have skip status")
-	require.Contains(t, skippedTest.Error.Error(), "This test is intentionally skipped", "Skip message should be preserved")
+	require.Nil(t, skippedTest.Error, "Skipped tests should not have error messages (skip reasons are not errors)")
```
Comment (janjakubnanista) [inline op-acceptor/runner/runner_test.go:1735]: This seems like a big change - going from having error to not having it - are we sure this ain't gonna break some downstreams?

**Review by janjakubnanista — CHANGES_REQUESTED** [review]
Some questions and small stuff, looks good otherwise!

**Review by janjakubnanista — APPROVED** [review]

File: op-acceptor/runner/adapters.go (hunk 5 above)
Comment (scharissis) [reply op-acceptor/runner/adapters.go]: I've gone ahead and removed this entirely to simplify things.
Cheers.

File: op-acceptor/runner/executor.go (hunk 8 above)
Comment (scharissis) [reply op-acceptor/runner/executor.go]: Agreed. I've refactored this to be error-based.
Cheers

File: op-acceptor/runner/jsonstore.go (hunk 9 above)
Comment (scharissis) [reply op-acceptor/runner/jsonstore.go:12]: Done

File: op-acceptor/runner/parser.go (hunk 10 above)
Comment (scharissis) [reply op-acceptor/runner/parser.go:17]: Done

File: op-acceptor/runner/parser.go (hunk 11 above)
Comment (scharissis) [reply op-acceptor/runner/parser.go]: Agreed. Extracted this to a function `isSubTestEvent`.
Cheers

File: op-acceptor/runner/runner_test.go (hunk 12 above)
Comment (scharissis) [reply op-acceptor/runner/runner_test.go:1735]: It's a change, but not a breaking change, as it's not exposed externally (to consumers).
Good call out, cheers!
